lib/newProfile.ml

```diff
@@ -141,23 +218,72 @@ let leave ?time name ?(args=[]) ?last () =
-     in the same span vs how much survived it *)
-  let major_words = mend.major_words -. mstart.major_words in
-  let minor_words = mend.minor_words -. mstart.minor_words in
-  let major_collect = mend.major_collections - mstart.major_collections in
-  let minor_collect = mend.minor_collections - mstart.minor_collections in
-  let ppf tdiff = `String (Format.sprintf "%.3G w" tdiff) in
-  let ppi i = `Intlit (string_of_int i) in
-  [("major_words",ppf major_words); ("minor_words", ppf minor_words);
-   ("major_collect", ppi major_collect); ("minor_collect", ppi minor_collect); ]
-
-let make_instr_diff ~istart ~iend =
-  match System.instructions_between ~c_start:istart ~c_end:iend with
-  | Ok count -> [("instr", `Intlit (Int64.to_string count))]
-  | Error _ -> []
+let init_poll () =
+  Runtime_events.start();
+
+  let cursor = Runtime_events.create_cursor None in
+
+  let to_time =
+    (* Timestamp.to_int64 is in nanoseconds
+       the runtime_events timestamps com from a monotonic clock,
+       so they are not directly comparable to the times from Unix.gettimeofday
+       (and the monotonic clock is AFAICT not exposed so we can't use it to replace gettimeofday).
+       Instead we calibrate by taking gettimeofday
+       then immediately generating an event to read its timestamp
+       and claim that the 2 are equal.
+    *)
+    let now = gettime() in
+    let now_ts = ref None in
+    let _ : int =
+      let runtime_begin _ ts _ = now_ts := Some ts in
+      Gc.minor (); (* ensure we have a recent event to look at *)
+      Runtime_events.(read_poll cursor (Callbacks.create ~runtime_begin ()) None)
+    in
+    let now_ts = Runtime_events.Timestamp.to_int64 (Option.get !now_ts) in
+
+    let open Int64 in
+    fun ts ->
+      now +. (to_float (sub (Runtime_events.Timestamp.to_int64 ts) now_ts) /. 1_000_000_000.)
+  in
+
+  let callbacks =
+    let open Runtime_events in
+    (* do we actually need a stack or is a start_of_current_phase ref enough? *)
+    let current_minor = ref [] in
+    let current_major = ref [] in
+    let runtime_begin _ ts = function
+      | EV_MINOR ->
+        current_minor := ts :: !current_minor;
+        duration ~is_gc:true ~time:(to_time ts) "gc_minor" "B" ()
+      | EV_MAJOR ->
+        current_major := ts :: !current_major;
+        duration ~is_gc:true ~time:(to_time ts) "gc_major" "B" ()
+      | _ -> ()
+    in
+    let update ref stack ts =
+      match !stack with
+      | [] -> assert false
+      | start :: rest ->
+        stack := rest;
+        ref := Int64.(add !ref (sub (Timestamp.to_int64 ts) (Timestamp.to_int64 start)))
+    in
+    let runtime_end _ ts = function
+      | EV_MINOR ->
+        update Counters.current_minor_time current_minor ts;
+        duration ~is_gc:true ~time:(to_time ts) "gc_minor" "E" ()
+      | EV_MAJOR ->
+        update Counters.current_major_time current_major ts;
+        duration ~is_gc:true ~time:(to_time ts) "gc_major" "E" ()
+      | _ -> ()
```

Comment:
NB: We could print for every event type but it doesn't seem particularly useful.